tokio/src/sync/rwlock.rs

```diff
@@ -77,19 +80,221 @@ pub struct RwLock<T> {
+where
+    T: fmt::Debug,
+{
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        fmt::Debug::fmt(&**self, f)
+    }
+}
+
+impl<'a, T> fmt::Display for MappedRwLockReadGuard<'a, T>
+where
+    T: fmt::Display,
+{
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        fmt::Display::fmt(&**self, f)
+    }
+}
+
+impl<'a, T> Drop for MappedRwLockReadGuard<'a, T> {
+    fn drop(&mut self) {
+        self.s.release(1);
+    }
+}
+
+/// An RAII write lock guard returned by [`RwLockWriteGuard::map`], which can
+/// point to a subfield of the protected data.
+///
+/// [`RwLockWriteGuard::map`]: method@RwLockWriteGuard::map
+pub struct MappedRwLockWriteGuard<'a, T> {
+    s: &'a Semaphore,
+    data: *mut T,
+    marker: marker::PhantomData<&'a mut T>,
+}
+
+impl<'a, T> ops::Deref for MappedRwLockWriteGuard<'a, T> {
+    type Target = T;
+
+    #[inline]
+    fn deref(&self) -> &T {
+        unsafe { &*self.data }
+    }
+}
+
+impl<'a, T> ops::DerefMut for MappedRwLockWriteGuard<'a, T> {
+    #[inline]
+    fn deref_mut(&mut self) -> &mut T {
+        unsafe { &mut *self.data }
+    }
+}
+
+impl<'a, T> fmt::Debug for MappedRwLockWriteGuard<'a, T>
+where
+    T: fmt::Debug,
+{
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        fmt::Debug::fmt(&**self, f)
+    }
+}
+
+impl<'a, T> fmt::Display for MappedRwLockWriteGuard<'a, T>
+where
+    T: fmt::Display,
+{
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        fmt::Display::fmt(&**self, f)
+    }
+}
+
+impl<'a, T> Drop for MappedRwLockWriteGuard<'a, T> {
+    fn drop(&mut self) {
+        self.s.release(MAX_READS);
+    }
+}
+
 /// RAII structure used to release the shared read access of a lock when
 /// dropped.
 ///
 /// This structure is created by the [`read`] method on
 /// [`RwLock`].
 ///
 /// [`read`]: method@RwLock::read
-#[derive(Debug)]
+/// [`RwLock`]: struct@RwLock
 pub struct RwLockReadGuard<'a, T> {
-    permit: ReleasingPermit<'a, T>,
     lock: &'a RwLock<T>,
 }
 
+impl<'a, T> RwLockReadGuard<'a, T> {
+    /// Make a new `MappedRwLockReadGuard` for a component of the locked data.
+    ///
+    /// This operation cannot fail as the `RwLockReadGuard` passed in already
+    /// locked the data.
+    ///
+    /// This is an associated function that needs to be
+    /// used as `RwLockReadGuard::map(...)`. A method would interfere with
+    /// methods of the same name on the contents of the locked data.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use tokio::sync::{RwLock, RwLockReadGuard};
+    ///
+    /// #[derive(Debug, Clone, Copy, PartialEq, Eq)]
+    /// struct Foo(u32);
+    ///
+    /// # #[tokio::main]
+    /// # async fn main() {
+    /// let lock = RwLock::new(Foo(1));
+    /// let mapped = RwLockReadGuard::map(lock.read().await, |f| &f.0);
+    /// assert_eq!(1, *mapped);
+    /// # }
+    /// ```
+    #[inline]
+    pub fn map<U, F>(self, f: F) -> MappedRwLockReadGuard<'a, U>
```

Comment:
Thanks for the pointer. My head just refactored the methods into `self` without realizing why.

> Also, are you planning on adding `map` and `try_map` associated functions on `MappedRwLockReadGuard` and `MappedRwLockWriteGuard`? Without that it will only be possible to map the data once.

Not at the moment since I personally have no use for it. Feel free to add it in a follow up to this.

Reply:
I was proof-reading documentation and decided to add associated methods for the mapped variants. Please look them over. Thanks.